backend/s3/s3.go

```diff
@@ -0,0 +1,200 @@
+package s3
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/mitchellh/goamz/aws"
+	as3 "github.com/mitchellh/goamz/s3"
+
+	"github.com/restic/restic/backend"
+)
+
+var ErrWrongData = errors.New("wrong data returned by backend, checksum does not match")
+
+func s3path(t backend.Type, name string) string {
+	if t == backend.Config {
+		return string(t)
+	}
+	return string(t) + "/" + name
+}
+
+type S3DB struct {
+	bucket *as3.Bucket
+	mput   sync.Mutex
+	path   string
+}
+
+// Open opens the local backend at bucket and region.
+func Open(regionname, bucketname string) (*S3DB, error) {
+	auth, err := aws.EnvAuth()
+	if err != nil {
+		return nil, err
+	}
+
+	client := as3.New(auth, aws.Regions[regionname])
+
+	return &S3DB{bucket: client.Bucket(bucketname), path: bucketname}, nil
+}
+
+// Location returns this backend's location (the directory name).
+func (b *S3DB) Location() string {
+	return b.path
+}
+
+type s3Blob struct {
+	b     *S3DB
+	buf   *bytes.Buffer
+	final bool
+}
+
+func (bb *s3Blob) Write(p []byte) (int, error) {
+	if bb.final {
+		return 0, errors.New("blob already closed")
+	}
+
+	n, err := bb.buf.Write(p)
+	return n, err
+}
+
+func (bb *s3Blob) Read(p []byte) (int, error) {
+	return bb.buf.Read(p)
+}
+
+func (bb *s3Blob) Close() error {
+	bb.final = true
+	bb.buf.Reset()
+	return nil
+}
+
+func (bb *s3Blob) Size() uint {
+	return uint(bb.buf.Len())
+}
+
+func (bb *s3Blob) Finalize(t backend.Type, name string) error {
+	if bb.final {
+		return errors.New("Already finalized")
+	}
+
+	bb.final = true
+
+	path := s3path(t, name)
+	bb.b.mput.Lock()
+	err := bb.b.bucket.Put(path, bb.buf.Bytes(), "binary/octet-stream", "private")
```

Comment:
I thinkt that's the implementation for "use just one connection". For using an arbitrary number, have a look at the `blobToken` channel in `archiver.go`. Basically, it's a channel with a limited buffer capacity which is filled with token (just an empty struct, `struct{}`). The code can then receive one token from the channel, which blocks if the channel is empty (aka: there are no token available at the moment), and put the token back once the goroutine is done.